logr.go

```diff
@@ -382,12 +477,90 @@ func (l Logger) IsZero() bool {
 	return l.sink == nil
 }
 
+// WithContext stores a context in the Logger. If the Logger also has keys set
+// that it is meant to log from a context, then the values for those keys from
+// the given context will be added to all log entries.
+func (l Logger) WithContext(ctx context.Context) Logger {
+	if l.sink == discardLogSinkInstance {
+		return l
+	}
+
+	l.setContext(ctx)
+	return l
+}
+
+// WithContextHandlers extends the list of context handlers. When
+// given a context through WithContext later, the Logger will invoke
+// each handler and log the key/value pairs that they return as
+// as if they had been passed to WithValues.
+func (l Logger) WithContextHandlers(handlers ...ContextHandler) Logger {
+	if l.sink == discardLogSinkInstance {
+		return l
+	}
+
+	l.setContextHandlers(handlers...)
+	return l
+}
+
+// mergeHandlers combines different slices into one. If a single one is passed,
+// it will still create a new slice and copy handlers to make the result
+// independent.
+func mergeHandlers(handlerSlices ...[]ContextHandler) []ContextHandler {
+	n := 0
+	for _, handlers := range handlerSlices {
+		n += len(handlers)
+	}
+	if n == 0 {
+		return nil
+	}
+	cp := make([]ContextHandler, 0, n)
+	for _, handlers := range handlerSlices {
+		cp = append(cp, handlers...)
+	}
+	return cp
+}
+
+func (l Logger) valuesFromContext() []interface{} {
+	if l.contextKV == nil || l.contextKV.ctx == nil || len(l.contextKV.handlers) == 0 {
+		return nil
+	}
+
+	// Fast path for a single handler: just return its result directly.
+	if len(l.contextKV.handlers) == 1 {
+		return l.contextKV.handlers[0].ValuesFromContext(l.contextKV.ctx)
+	}
+
+	// We don't really know how large this has to be. Let's
+	// assume that each handler returns at most two key/value
+	// pairs, of which each needs two entries.
+	keysAndValues := make([]interface{}, 0, 64*len(l.contextKV.handlers))
+	for _, handler := range l.contextKV.handlers {
+		keysAndValues = append(keysAndValues, handler.ValuesFromContext(l.contextKV.ctx)...)
+	}
+	return keysAndValues
+}
+
+// ContextHandler provides callbacks for working with a context.
+// More functionality might get added in the future.
+type ContextHandler struct {
+	// ValuesFromContext returns additional key/value pairs for data that
+	// was stored in the context. These pairs will be logged via WithValues
+	// and thus may get overridden by explicit log parameters if there is a
+	// key conflict.
+	//
+	// ValuesFromContext must always return the same key/value pairs for
+	// the same context. The result might get cached.
+	ValuesFromContext func(ctx context.Context) []interface{}
```

Comment:
Naming: "context" is redunant.  I am not 100% convinced this struct is necessary.  Wouldn't it all be easier to just say

```
log = log.WithContextValues(extractFromContext1, extractFromContext2)
```

Reply:
I wanted to avoid adding new APIs when at some point we find that we want to use something else then `func(ctx context.Context) []interface{}`. But I suppose I'm just overly cautious here: if that need ever arises, we can still introduce such a struct internally and have two APIs for adding to it (`WithContextValues` and `WithContextValuesViaXYZ` or whatever the difference will be).

I'm switching to `WithContextValues` taking `ContextReader = func(ctx context.Context) []interface{}` as arguments.